### PR TITLE
[Experimental] Reliability fixes for new filter blocks

### PR DESCRIPTION
**Pull request by dinhtungdu**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR is extracted from #51624 while working on #51622. This PR includes various bug fixes to improve the stability and reliability of new filter blocks powered by interactivity API. I break the original PR down into two separate PRs to make them easier to review.

This PR:

- adds the `getInnerBlockBy` function enabling retrieving inner blocks by callback.
- updates the naming convention for the target attribute key when querying element value.
- removes unused `clear-button/frontend.ts` file.
- updates block context `filterData` type.
- adds the `data-wc-key` attribute to each filter block to prevent the diffing issue.
- fixes the styling issues after navigation by always rendering all inner blocks (even the block doesn't have any filter options) and hiding the block using the `hidden` attribute.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

0. Ensure the experimental blocks are enabled:
    - If you use the WooCommerce Beta Tester plugin, you can enable the experimental blocks in Tools > WCA Test Helper > Features. Toggle the `experimental-blocks` option.
    - If you build this PR locally, build the branch with the `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` command to have experimental blocks available.
1.  Ensure the `Product Filters` template part is reset.
2. Add another Attribute filter block to the template part, and select a different attribute.
3. Inside that new attribute filter block, add a block that doesn't exist on the page, for example, Cover block.
4. Add the `Product Filters (Experimental)` template part to the `Product Catalog` template.
5. Save and go to the shop page.
6. Select an attribute, see the URL updated and a clear button appears.
7. Click on the clear button clears the selected filters.
8. On the first attribute filter block that doesn't contain the cover block, select an attribute that make the second attribute filter block disappear.
9. Reload the page, see the selected filter persists and the second attribute filter block is still hidden.
10. Clear the selected filter, and see the second filter block appear with the cover block style is loaded correctly.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/attribute-filter/frontend.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ store( 'woocommerce/product-filter-attribute', {
 		toggleFilter: () => {
 			const { ref } = getElement();
 			const targetAttribute =
-				ref.getAttribute( 'data-attribute-value' ) ?? 'value';
+				ref.getAttribute( 'data-target-value' ) ?? 'value';
```
Comment (gigitux) [inline plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/attribute-filter/frontend.ts:22]: is `ref.getAttribute( 'data-target-value' )` necessary?

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterCheckboxList.php [hunk 2]
```diff
@@ -24,10 +24,14 @@ final class ProductFilterCheckboxList extends AbstractBlock {
 	 * @return string Rendered block type output.
 	 */
 	protected function render( $attributes, $content, $block ) {
+		if ( empty( $block->context['filterData'] ) || empty( $block->context['filterData']['items'] ) ) {
+			return '';
```
Comment (gigitux) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterCheckboxList.php:28]: Could it be useful adding a warning log here?

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterChips.php [hunk 3]
```diff
@@ -24,12 +24,15 @@ final class ProductFilterChips extends AbstractBlock {
 	 * @return string Rendered block type output.
 	 */
 	protected function render( $attributes, $content, $block ) {
+		if ( empty( $block->context['filterData'] ) || empty( $block->context['filterData']['items'] ) ) {
+			return '';
```
Comment (gigitux) [inline plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterChips.php:28]: Could it be useful adding a warning log here?

**Review by gigitux — APPROVED** [review]
Thanks for working on this, @dinhtungdu! Brilliant work! 

I explored this code while I worked on https://github.com/woocommerce/woocommerce/pull/51714, so I thought that it was a good idea to review this PR. The PR LGTM! I left some questions/hints, but feel free to merge this PR as it is!

**Review by thealexandrelara — APPROVED** [review]
Really nice, thank you for working on improving the block reliability!

## Test summary:

✅ Scenario 1: The Product Filters template part correctly renders on the editor
- Conditions:
    - A merchant visits the Product Filters (Experimental) template part.
- Expected Outcome: The template part and its inner blocks are correctly rendered.
- Details: Everything is being rendered as expected.

---

✅ Scenario 2: Shoppers can add extra blocks to the Attribute filter block
- Conditions:
    - A merchant visits the Product Filters (Experimental) template part.
    - The merchant adds a second Attribute filter block that uses a different attribute than the default one.
    - The merchant adds an extra block, such as the Cover block, within the newly added Attribute filter block.
    - The merchant saves the template part.
    - The merchant visits the Product Catalog template, adds the template part to it, and saves the template.
    - A shopper visits the Shop page (/shop).
- Expected Outcome: The second Attribute filter block should contain the added block.
- Details: When visiting the Shop page I can confirm that the added block is being correctly rendered and displayed on the frontend.

---

✅ Scenario 3: Attribute Filter blocks with extra blocks added to it are completely hidden when there is no products matching those attributes.
- Conditions:
    - A merchant visits the Product Filters (Experimental) template part.
    - The merchant adds a second Attribute filter block that uses a different attribute than the default one.
    - The merchant adds an extra block, such as the Cover block, within the newly added Attribute filter block.
    - The merchant saves the template part.
    - The merchant visits the Product Catalog template, adds the template part to it, and saves the template.
    - A shopper visits the Shop page (/shop).
    - The shopper selects a set of filters that results in an state where no products are available for the second added Attribute filter.
- Expected Outcome: The second Attribute filter block should be completely hidden including its inner blocks.
- Details: When the selected filters cause an state where the second Attribute filter does not have any products with those attributes, then the block is completely hidden as expected.

File: plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/attribute-filter/frontend.ts (hunk 1 above)
Comment (dinhtungdu) [reply plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/attribute-filter/frontend.ts:22]: It allows you to customize the attribute key that holds the value. For example, for some reason, you don't want to use the `value` attribute, you can do that with data-target-value.

File: plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterCheckboxList.php (hunk 2 above)
Comment (dinhtungdu) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterCheckboxList.php:28]: Can you give an example of the benefit of a warning here? And what type of warning are you thinking about? I don't think you meant [a PHP warning](https://www.php.net/manual/en/function.trigger-error.php).

Comment (gigitux) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterCheckboxList.php:28]: When working on [this pull request](https://github.com/woocommerce/woocommerce/pull/51714), I noticed that the block wasn’t displayed at all. After some code digging, I realized that I wasn't correctly passing the filter data.

This block could potentially be used by third-party developers as well. Including a PHP warning that indicates when filterData isn’t passed to the block could be useful for debugging purposes.

Comment (gigitux) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterCheckboxList.php:28]: As I said above, mine isn't a strong opinion. Feel free to merge the PR as it is! Amazing work! 🙇

Comment (dinhtungdu) [reply plugins/woocommerce/src/Blocks/BlockTypes/ProductFilterCheckboxList.php:28]: > This block could potentially be used by third-party developers as well.

In my design, it's only for filter blocks. Chips and Checkbox List aren't meant to be extended or reused, they're designed only for filter blocks.

I will think more about it though, as we have three types of audience for the warning: developers, merchants/admins, and shoppers/visitors. I also don't have a good idea right now on how to do it. This PR is blocking some other PR so let's merge it and implement the messaging later if needed.

File: plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/attribute-filter/frontend.ts (hunk 1 above)
Comment (gigitux) [reply plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/attribute-filter/frontend.ts:22]: Fixed with 3ad9e9a107e8d344471c4da2d4cac49e3aaf0b1c

**Pull request merged**